File: src/main/java/com/hepdd/gtmthings/api/misc/UnlimitedItemStackTransfer.java
package com.hepdd.gtmthings.api.misc;

import com.gregtechceu.gtceu.api.transfer.item.CustomItemStackHandler;

import net.minecraft.core.NonNullList;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.Tag;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.items.ItemHandlerHelper;

import lombok.Setter;
import org.jetbrains.annotations.NotNull;

import java.util.function.Predicate;

import javax.annotation.Nonnull;

public class UnlimitedItemStackTransfer extends CustomItemStackHandler {

    public UnlimitedItemStackTransfer(int size) {
        super(size);
    }

    public UnlimitedItemStackTransfer(NonNullList<ItemStack> stacks) {
        super(stacks);
    }

    public UnlimitedItemStackTransfer(ItemStack stack) {
        super(stack);
    }

    @Setter
    private Predicate<ItemStack> filter;

    @Override
    public boolean isItemValid(int slot, @Nonnull ItemStack stack) {
        return filter == null || filter.test(stack);
    }

    @Override
    @NotNull
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        if (amount == 0) return ItemStack.EMPTY;
        ItemStack existing = this.stacks[slot];
        int count = existing.getCount();
        if (count < 1) return ItemStack.EMPTY;
        if (count <= amount) {
            if (!simulate) {
                this.stacks[slot] = ItemStack.EMPTY;
                onContentsChanged(slot);
                return existing;
            } else {
                return existing.copy();
            }
        } else {
            if (!simulate) {
                existing.setCount(count - amount);
                onContentsChanged(slot);
            }
            return ItemHandlerHelper.copyStackWithSize(existing, amount);
        }
    }

    @Override
    public int getSlotLimit(int slot) {
        return Integer.MAX_VALUE;
    }

    @Override
    protected int getStackLimit(int slot, @NotNull ItemStack stack) {
        return Integer.MAX_VALUE;
    }

    @Override
    public CompoundTag serializeNBT() {
        ListTag nbtTagList = new ListTag();
        for (int i = 0; i < size; i++) {
            if (!stacks[i].isEmpty()) {
                CompoundTag itemTag = new CompoundTag();
                itemTag.putInt("Slot", i);
                var is = stacks[i].copy();
                itemTag.putInt("realCount", is.getCount());
                is.setCount(1);
                is.save(itemTag);
                nbtTagList.add(itemTag);
            }
        }
        CompoundTag nbt = new CompoundTag();
        nbt.put("Items", nbtTagList);
        nbt.putInt("Size", size);
        return nbt;
    }

    @Override
    public void deserializeNBT(CompoundTag nbt) {
        setSize(nbt.contains("Size", Tag.TAG_INT) ? nbt.getInt("Size") : size);
        ListTag tagList = nbt.getList("Items", Tag.TAG_COMPOUND);
        for (int i = 0; i < tagList.size(); i++) {
            CompoundTag itemTags = tagList.getCompound(i);
            int slot = itemTags.getInt("Slot");

            if (slot >= 0 && slot < size) {
                var is = ItemStack.of(itemTags).copy();
                is.setCount(itemTags.getInt("realCount"));
                stacks[slot] = is;
            }
        }
    }
}

File: src/main/java/com/hepdd/gtmthings/api/misc/UnlimitedItemStackTransfer.kt
package com.hepdd.gtmthings.api.misc

import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.nbt.Tag
import net.minecraft.world.item.ItemStack

import com.gregtechceu.gtceu.api.transfer.item.CustomItemStackHandler
import com.lowdragmc.lowdraglib.side.item.ItemTransferHelper

import java.util.function.Function
import kotlin.math.min

class UnlimitedItemStackTransfer(size: Int) : CustomItemStackHandler(size) {

    private val filter: Function<ItemStack?, Boolean?>? = null

    override fun isItemValid(slot: Int, stack: ItemStack): Boolean = filter == null || filter.apply(stack) == true

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        if (amount == 0) return ItemStack.EMPTY

        validateSlotIndex(slot)

        val existing = this.stacks[slot]

        if (existing.isEmpty) return ItemStack.EMPTY

        val toExtract = min(amount, getSlotLimit(slot))

        if (existing.count <= toExtract) {
            if (!simulate) {
                this.stacks[slot] = ItemStack.EMPTY
                onContentsChanged(slot)
                return existing
            } else {
                return existing.copy()
            }
        } else {
            if (!simulate) {
                this.stacks[slot] = ItemTransferHelper.copyStackWithSize(existing, existing.count - toExtract)
                onContentsChanged(slot)
            }

            return ItemTransferHelper.copyStackWithSize(existing, toExtract)
        }
    }

    override fun getSlotLimit(slot: Int): Int = Int.Companion.MAX_VALUE

    override fun getStackLimit(slot: Int, stack: ItemStack): Int = Int.Companion.MAX_VALUE

    override fun serializeNBT(): CompoundTag {
        val nbtTagList = ListTag()
        for (i in stacks.indices) {
            if (!stacks[i].isEmpty) {
                val itemTag = CompoundTag()
                itemTag.putInt("Slot", i)
                val `is` = stacks[i].copy()
                itemTag.putInt("realCount", `is`.count)
                `is`.count = 1
                `is`.save(itemTag)
                nbtTagList.add(itemTag)
            }
        }
        val nbt = CompoundTag()
        nbt.put("Items", nbtTagList)
        nbt.putInt("Size", stacks.size)
        return nbt
    }

    override fun deserializeNBT(nbt: CompoundTag) {
        setSize(if (nbt.contains("Size", Tag.TAG_INT.toInt())) nbt.getInt("Size") else stacks.size)
        val tagList = nbt.getList("Items", Tag.TAG_COMPOUND.toInt())
        for (i in tagList.indices) {
            val itemTags = tagList.getCompound(i)
            val slot = itemTags.getInt("Slot")

            if (slot >= 0 && slot < stacks.size) {
                val `is` = ItemStack.of(itemTags).copy()
                `is`.count = itemTags.getInt("realCount")
                stacks[slot] = `is`
            }
        }
        onLoad()
    }
}
